### PR TITLE
Bulk Load CDK: Staging refactor + tests; multi-sync ITs correctly wait for ack

### PR DESCRIPTION
**Pull request by johnny-schmidt**
## What
First three commits are modifications to state so that it is answering specific questions rather than being inspected
Also tests around ensuring those questions are answered

Next two move staging from processBatch to close and disable the tests that are incorrectly not awaiting a checkpoint ack

Next 6 are @edgao 's test fixes to make the multi-sync tests wait on ack

Last one tweaks that to run more efficiently and not use filler records; it also ensures that destination state is persisted after each file is written to staging, so that orphaned staged data after a failure is recovered

**EDIT**
Plus
* one commit I accidentally stranded that makes staging fall back to metadata when the file isn't present
* a couple of attempts to keep mock dest it from breaking

### CODE REVIEW

**Comment by johnny-schmidt** [comment]
fixed the failing mock dest IT + rebased

**Review by edgao — APPROVED** [review]
tests failing in timeout - that's probably expected, given that we're pushing huge data per test. Could try bumping the timeout in `destination-s3-v2/gradle.properties`, there should be examples in some other destination

(or reduce the number of messages, if we're flushing on every message)

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/state/SyncManager.kt [hunk 1]
```diff
@@ -79,7 +80,8 @@ class DefaultSyncManager(
         stream: DestinationStream.Descriptor
     ): StreamLoader? {
         val completable = streamLoaders[stream]
-        return completable?.let { if (it.isCompleted) it.await() else null }
+        // `.isCompleted` does not work as expected here.
```
[the remark this answers is not in the document]
Comment (johnny-schmidt) [reply airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/state/SyncManager.kt:83]: It did not return true sometimes even when the loader was clearly completed.

File: airbyte-cdk/bulk/core/load/src/testFixtures/resources/application.yaml [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+    enabled: ${USE_FILE_TRANSFER:false}
+    staging-path: ${AIRBYTE_STAGING_DIRECTORY:/staging/files}
+  resources:
+    disk:
+      bytes: ${CONNECTOR_STORAGE_LIMIT_BYTES:5368709120} # 5GB
+  flush:
+    rate-ms: 900000 # 15 minutes
+    window-ms: 900000 # 15 minutes
+  destination:
+    record-batch-size: 1 # 1 byte for testing; 1 record => 1 upload
```
[the remark this answers is not in the document]
Comment (johnny-schmidt) [reply airbyte-cdk/bulk/core/load/src/testFixtures/resources/application.yaml:12]: I disabled that?

[the remark this answers is not in the document]
Comment (edgao) [reply airbyte-cdk/bulk/core/load/src/testFixtures/resources/application.yaml:12]: ah derp, I didn't read that file b/c I thought it was just my diff. this makes sense!

(... though I'm not sure why we're hitting test timeouts then)

**Comment by johnny-schmidt** [comment]
@edgao The docker tests were timing out because they needed to consume the batch size limitation via an env variable. I hacked it in. Also they don't throw unclean exit on kill, so I hacked in an exception.

Very hack all the way, but they should pass now.

**Pull request merged**